File: client/cli/src/submit.rs
//! Submit command implementation for the Pie CLI.
//!
//! This module implements the `pie submit` subcommand for submitting inferlets
//! to an existing running Pie engine instance.

use crate::engine;
use crate::path;
use anyhow::Context;
use anyhow::Result;
use clap::Args;
use pie_client::client;
use std::fs;
use std::path::PathBuf;
use wac_graph::{CompositionGraph, EncodeOptions};
use wac_types::Package;

/// Arguments for the `pie submit` command.
#[derive(Args, Debug)]
pub struct SubmitArgs {
    /// Path to the .wasm inferlet file.
    #[arg(value_parser = path::expand_tilde)]
    pub inferlet: PathBuf,
    /// Path to a custom TOML configuration file.
    #[arg(long)]
    pub config: Option<PathBuf>,
    /// The network host to connect to.
    #[arg(long)]
    pub host: Option<String>,
    /// The network port to connect to.
    #[arg(long)]
    pub port: Option<u16>,
    /// The username to use for authentication.
    #[arg(long)]
    pub username: Option<String>,
    /// Path to the private key file to use for authentication.
    #[arg(long)]
    pub private_key_path: Option<PathBuf>,
    /// Run the inferlet in detached mode.
    #[arg(short, long, default_value = "false")]
    pub detached: bool,
    /// Paths to .wasm library files to link with the inferlet.
    #[arg(short, long, value_parser = path::expand_tilde)]
    pub link: Vec<PathBuf>,
    /// Arguments to pass to the inferlet after `--`.
    #[arg(last = true)]
    pub arguments: Vec<String>,
}

/// Compose a component with a library. The library (plug) is linked into the
/// main program (socket).
fn compose_one_component(socket_bytes: Vec<u8>, plug_bytes: Vec<u8>) -> Result<Vec<u8>> {
    // Create a new composition graph
    let mut graph = CompositionGraph::new();

    // Prepare the socket package
    let socket_package = Package::from_bytes("socket", None, socket_bytes, graph.types_mut())
        .context("Failed to load socket component")?;
    let socket_id = graph
        .register_package(socket_package)
        .context("Failed to register socket package")?;

    // Prepare the plug package
    let plug_package = Package::from_bytes("plug", None, plug_bytes, graph.types_mut())
        .context("Failed to load plug component")?;
    let plug_id = graph
        .register_package(plug_package)
        .context("Failed to register plug package")?;

    // Perform the composition: connect the plug's exports to the socket's imports
    wac_graph::plug(&mut graph, vec![plug_id], socket_id)
        .context("Failed to compose components")?;

    // Encode the composed component to bytes
    let composed_bytes = graph
        .encode(EncodeOptions::default())
        .context("Failed to encode composed component")?;

    Ok(composed_bytes)
}

/// Compose a component with multiple libraries. The libraries are linked into the main program
/// sequentially in the order of the library paths.
///
/// Notably, the composition is done iteratively. At each iteration, a library is linked into
/// the main program. This allows an interface instrumented by a library to be instrumented
/// again by another following library.
fn compose_components(program_bytes: Vec<u8>, library_paths: &[PathBuf]) -> Result<Vec<u8>> {
    let mut socket_bytes = program_bytes;

    for library_path in library_paths {
        let plug_bytes = fs::read(library_path)
            .context(format!("Failed to read library file at {:?}", library_path))?;
        let composed_bytes = compose_one_component(socket_bytes, plug_bytes)?;
        socket_bytes = composed_bytes;
    }

    Ok(socket_bytes)
}

/// Handles the `pie submit` command.
///
/// This function:
/// 1. Reads configuration from the specified config file or default config
/// 2. Creates a client configuration from config and command-line arguments
/// 3. Connects to the existing Pie engine server
/// 4. If libraries are specified, composes them with the inferlet on the client side
/// 5. Submits the composed inferlet with the provided arguments
/// 6. In non-detached mode, streams the inferlet output with signal handling:
///    - Ctrl-C (SIGINT): Terminates the inferlet on the server
///    - Ctrl-D (EOF): Detaches from the inferlet (continues running on server)
pub async fn handle_submit_command(args: SubmitArgs) -> Result<()> {
    let client_config = engine::ClientConfig::new(
        args.config,
        args.host,
        args.port,
        args.username,
        args.private_key_path,
    )?;

    let client = engine::connect_and_authenticate(&client_config).await?;

    // Read the main inferlet
    let inferlet_blob = fs::read(&args.inferlet)
        .context(format!("Failed to read Wasm file at {:?}", args.inferlet))?;

    // If libraries are specified, compose them with the main inferlet
    let final_blob = if args.link.is_empty() {
        inferlet_blob
    } else {
        compose_components(inferlet_blob, &args.link)
            .context("Failed to compose inferlet with libraries")?
    };

    // Calculate the hash of the final composed blob
    let hash = client::hash_blob(&final_blob);
    println!("Final inferlet hash: {}", hash);

    // Upload the composed inferlet to the server
    if !client.program_exists(&hash).await? {
        client.upload_program(&final_blob).await?;
        println!("✅ Inferlet upload successful.");
    } else {
        println!("Inferlet already exists on server.");
    }

    let cmd_name = args
        .inferlet
        .file_stem()
        .context("Inferlet path must have a valid file name")?
        .to_string_lossy()
        .to_string();

    let instance = client
        .launch_instance(hash, cmd_name, args.arguments, args.detached)
        .await?;

    println!("✅ Inferlet launched with ID: {}", instance.id());

    if !args.detached {
        engine::stream_inferlet_output(instance, client).await?;
    }

    Ok(())
}

File: client/cli/src/pie_client_cli/submit.py
"""Submit command implementation for the Pie CLI.

This module implements the `pie-cli submit` subcommand for submitting inferlets
to an existing running Pie engine instance.
"""

import subprocess
import tempfile
from pathlib import Path
from typing import Optional

import blake3

from . import engine


def compose_components(program_bytes: bytes, library_paths: list[Path]) -> bytes:
    """Compose a program with multiple libraries using wac CLI.
    
    Uses `wac plug` command to link libraries into the main program.
    Libraries are linked sequentially in the order provided.
    
    Args:
        program_bytes: The main program WASM bytes.
        library_paths: List of paths to library WASM files.
    
    Returns:
        The composed WASM bytes.
    
    Raises:
        RuntimeError: If wac CLI is not available or composition fails.
    """
    # Check if wac is available
    try:
        result = subprocess.run(
            ["wac", "--version"],
            capture_output=True,
            text=True,
        )
        if result.returncode != 0:
            raise RuntimeError("wac CLI is not available")
    except FileNotFoundError:
        raise RuntimeError(
            "wac CLI is not installed. Install it with: cargo install wac-cli"
        )
    
    socket_bytes = program_bytes
    
    for library_path in library_paths:
        if not library_path.exists():
            raise FileNotFoundError(f"Library file not found: {library_path}")
        
        # Read library bytes
        plug_bytes = library_path.read_bytes()
        
        # Compose using wac plug
        # wac plug --plug <library.wasm> <main.wasm> -o <output.wasm>
        with tempfile.TemporaryDirectory() as temp_dir:
            temp_path = Path(temp_dir)
            socket_file = temp_path / "socket.wasm"
            plug_file = temp_path / "plug.wasm"
            output_file = temp_path / "composed.wasm"
            
            socket_file.write_bytes(socket_bytes)
            plug_file.write_bytes(plug_bytes)
            
            result = subprocess.run(
                [
                    "wac", "plug",
                    "--plug", str(plug_file),
                    str(socket_file),
                    "-o", str(output_file),
                ],
                capture_output=True,
                text=True,
            )
            
            if result.returncode != 0:
                raise RuntimeError(
                    f"wac plug failed for {library_path}:\n{result.stderr}"
                )
            
            socket_bytes = output_file.read_bytes()
    
    return socket_bytes


def handle_submit_command(
    inferlet: Path,
    config: Optional[Path] = None,
    host: Optional[str] = None,
    port: Optional[int] = None,
    username: Optional[str] = None,
    private_key_path: Optional[Path] = None,
    detached: bool = False,
    link: Optional[list[Path]] = None,
    arguments: Optional[list[str]] = None,
) -> None:
    """Handle the `pie-cli submit` command.
    
    1. Creates a client configuration from config file and command-line arguments
    2. Connects to the Pie engine server
    3. If libraries are specified, composes them with the inferlet using wac
    4. Uploads the composed inferlet if not already on server
    5. Launches the inferlet with the provided arguments
    6. In non-detached mode, streams the inferlet output with signal handling
    """
    link = link or []
    arguments = arguments or []
    
    client_config = engine.ClientConfig.create(
        config_path=config,
        host=host,
        port=port,
        username=username,
        private_key_path=private_key_path,
    )
    
    client = engine.connect_and_authenticate(client_config)
    
    try:
        # Read the main inferlet
        if not inferlet.exists():
            raise FileNotFoundError(f"Inferlet file not found: {inferlet}")
        
        inferlet_blob = inferlet.read_bytes()
        
        # If libraries are specified, compose them with the main inferlet
        if link:
            final_blob = compose_components(inferlet_blob, link)
        else:
            final_blob = inferlet_blob
        
        # Calculate the hash of the final composed blob
        program_hash = blake3.blake3(final_blob).hexdigest()
        print(f"Final inferlet hash: {program_hash}")
        
        # Upload the composed inferlet to the server
        if not engine.program_exists(client, program_hash):
            engine.upload_program(client, final_blob)
            print("✅ Inferlet upload successful.")
        else:
            print("Inferlet already exists on server.")
        
        # Get command name from inferlet filename
        cmd_name = inferlet.stem
        
        # Launch the instance
        instance = engine.launch_instance(
            client,
            program_hash,
            cmd_name,
            arguments,
            detached,
        )
        
        print(f"✅ Inferlet launched with ID: {instance.instance_id}")
        
        if not detached:
            engine.stream_inferlet_output(instance, client)
    
    finally:
        engine.close_client(client)
